crates/services/p2p/src/cached_view.rs

```diff
@@ -0,0 +1,319 @@
+        self.sealed_block_headers.clear();
+        self.transactions_on_blocks.clear();
+    }
+
+    fn update_metrics<U>(&self, update_fn: U)
+    where
+        U: FnOnce(),
+    {
+        if self.metrics {
+            update_fn()
+        }
+    }
+
+    pub(crate) fn get_sealed_headers<V>(
+        &self,
+        view: &V,
+        block_height_range: Range<u32>,
+    ) -> StorageResult<Option<Vec<SealedBlockHeader>>>
+    where
+        V: P2pDb,
+    {
+        let mut headers = Vec::new();
+        let mut missing_ranges = block_height_range.clone();
+        for block_height in block_height_range.clone() {
+            if let Some(header) = self.sealed_block_headers.get(&block_height) {
+                headers.push(header.clone());
+            } else {
+                // for the first block not in the cache, start a new range
+                missing_ranges.start = block_height;
+                break;
+            }
+        }
+
+        if missing_ranges.is_empty() {
+            self.update_metrics(increment_p2p_req_res_cache_hits);
+            return Ok(Some(headers))
+        }
+
+        self.update_metrics(increment_p2p_req_res_cache_misses);
+        let missing_headers = view.get_sealed_headers(missing_ranges.clone())?;
+        if let Some(missing_headers) = &missing_headers {
+            for header in missing_headers.iter() {
+                self.sealed_block_headers
+                    .insert((*header.entity.height()).into(), header.clone());
+                headers.push(header.clone());
+            }
+        }
+        Ok(missing_headers)
+    }
+
+    pub(crate) fn get_transactions<V>(
+        &self,
+        view: &V,
+        block_height_range: Range<u32>,
+    ) -> StorageResult<Option<Vec<Transactions>>>
+    where
+        V: P2pDb,
+    {
+        let mut transactions = Vec::new();
+        let mut missing_ranges = block_height_range.clone();
+        for block_height in block_height_range.clone() {
+            if let Some(cached_tx) = self.transactions_on_blocks.get(&block_height) {
+                transactions.push(cached_tx.clone());
+            } else {
+                // for the first block not in the cache, start a new range
+                missing_ranges.start = block_height;
+                break;
+            }
+        }
+
+        if missing_ranges.is_empty() {
+            self.update_metrics(increment_p2p_req_res_cache_hits);
+            return Ok(Some(transactions))
+        }
+
+        self.update_metrics(increment_p2p_req_res_cache_misses);
+        let transactions = view.get_transactions(missing_ranges.clone())?;
+        if let Some(transactions) = &transactions {
+            for (block_height, transactions_per_block) in
+                missing_ranges.zip(transactions.iter())
+            {
+                self.transactions_on_blocks
+                    .insert(block_height, transactions_per_block.clone());
+            }
+        }
+        Ok(transactions)
```

Comment:
same here, it looks like only the transactions fetched because of a cache miss are returned?

Reply:
addressed in https://github.com/FuelLabs/fuel-core/commit/5ad509354d91f94dcf56713f71350138905819f0